config/confighttp/compressor.go

```diff
@@ -22,36 +22,32 @@ type writeCloserReset interface {
 	Reset(w io.Writer)
 }
 
-var (
-	_          writeCloserReset = (*gzip.Writer)(nil)
-	gZipPool                    = &compressor{pool: sync.Pool{New: func() any { return gzip.NewWriter(nil) }}}
-	_          writeCloserReset = (*snappy.Writer)(nil)
-	snappyPool                  = &compressor{pool: sync.Pool{New: func() any { return snappy.NewBufferedWriter(nil) }}}
-	_          writeCloserReset = (*zstd.Encoder)(nil)
-	// Concurrency 1 disables async decoding via goroutines. This is useful to reduce memory usage and isn't a bottleneck for compression using sync.Pool.
-	zStdPool                  = &compressor{pool: sync.Pool{New: func() any { zw, _ := zstd.NewWriter(nil, zstd.WithEncoderConcurrency(1)); return zw }}}
-	_        writeCloserReset = (*zlib.Writer)(nil)
-	zLibPool                  = &compressor{pool: sync.Pool{New: func() any { return zlib.NewWriter(nil) }}}
-)
-
 type compressor struct {
 	pool sync.Pool
 }
 
 // writerFactory defines writer field in CompressRoundTripper.
 // The validity of input is already checked when NewCompressRoundTripper was called in confighttp,
-func newCompressor(compressionType configcompression.Type) (*compressor, error) {
-	switch compressionType {
+func newCompressor(compressionType configcompression.TypeWithLevel) (*compressor, error) {
+	// compressionType, compressionLevel := getCompression(compressionType)
+
+	switch compressionType.Type {
 	case configcompression.TypeGzip:
-		return gZipPool, nil
+		var _ writeCloserReset = (*gzip.Writer)(nil)
+		return &compressor{pool: sync.Pool{New: func() any { w, _ := gzip.NewWriterLevel(nil, int(compressionType.Level)); return w }}}, nil
```

Comment:
The pools have to be kept reusable. We should not create them on every request

Reply:
yeah this one seems tricky, I moved the pools inside the function, because we now need to pass in the level to the writer. Going back this [comment](https://github.com/open-telemetry/opentelemetry-collector/issues/8216#issuecomment-1735802401). Do you think we need to pool the encoders in `confighttp`?

Reply:
@dmitryax I have made changes to keep the pools reusable as suggested. Please let me know if you see issues with it.